P178_Frogmore_SCD30_Arduino_Library/FrogmoreScd30.cpp

```diff
@@ -0,0 +1,653 @@
+/*
+#  Copyright (c) 2019 Frogmore42
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*/
+#include <Wire.h>
+#include <math.h>
+#include <stdio.h>
+#include <twi.h>
+#include <FrogmoreScd30.h>
+
+#define COMMAND_SCD30_CONTINUOUS_MEASUREMENT      0x0010
+#define COMMAND_SCD30_MEASUREMENT_INTERVAL        0x4600
+#define COMMAND_SCD30_GET_DATA_READY              0x0202
+#define COMMAND_SCD30_READ_MEASUREMENT            0x0300
+#define COMMAND_SCD30_CALIBRATION_TYPE            0x5306
+#define COMMAND_SCD30_FORCED_RECALIBRATION_FACTOR 0x5204
+#define COMMAND_SCD30_TEMPERATURE_OFFSET          0x5403
+#define COMMAND_SCD30_ALTITUDE_COMPENSATION       0x5102
+#define COMMAND_SCD30_SOFT_RESET                  0xD304
+#define COMMAND_SCD30_GET_FW_VERSION              0xD100
+#define COMMAND_SCD30_STOP_MEASUREMENT            0x0104
+
+#define SCD30_DATA_REGISTER_BYTES 2
+#define SCD30_DATA_REGISTER_WITH_CRC 3
+#define SCD30_MEAS_BYTES 18
+
+#ifdef SCD30_DEBUG
+enum LoggingLevels {LOG_LEVEL_NONE, LOG_LEVEL_ERROR, LOG_LEVEL_INFO, LOG_LEVEL_DEBUG, LOG_LEVEL_DEBUG_MORE, LOG_LEVEL_ALL};
+char scd30log_data[180];
+#endif
+
+void FrogmoreScd30::begin(TwoWire *pWire, uint8_t i2cAddress)
+{
+    this->i2cAddress = i2cAddress;
+    if (pWire == NULL)
+    {
+        this->pWire = &Wire;
+    }
+    else
+    {
+        this->pWire = pWire;
+    }
+
+    co2NewDataLocation = -1; // indicates there is no data, so the 1st data point needs to fill up the median filter
+    this->pWire->setClockStretchLimit(200000);
```

Comment:
A plugin or a library should not set parameters (or call `begin`, which you may also see sometimes) on a global object like I2C Wire.
It will affect all other devices.

If it is really needed to operate, you should let the user set it and make a note of it in the config page.
ESPeasy has a global option to set the clock stretch limit.

Reply:
OK, I just preferred to avoid modification of original code.